### ISSUE TITLE
Cleaning up CMOR input file vs registered content sources

### ISSUE THREAD

**Issue by durack1** [PCMDI/cmor#628]
@mauzey1 I have been trying to simplify my [CMOR_input.json](https://github.com/PCMDI/amipbcs/blob/master/CMOR/drive_input4MIPs_obs.json) file along with the details (that are currently duplicated) in the registered content, found in the [input4MIPs_CV.json](https://github.com/PCMDI/input4MIPs-cmor-tables/blob/issue87_durack1_PCMDI-AMIP-1-2-0AndPy3Update/Tables/input4MIPs_CV.json#L974-L1007) file.

My aspiration, was that all information that is contained in the registered content (`input4MIPs_CV.json`) would be the primary source, and only additional/runtime info e.g.

![Screen Shot 2021-09-02 at 1 28 28 PM](https://user-images.githubusercontent.com/3229632/131911725-aafb658b-82cc-46b6-8914-bf8583a17d74.png)
[link](https://github.com/PCMDI/amipbcs/blob/ac641dfa1bee45d4b259e891f06ca7e98b8734ea/CMOR/drive_input4MIPs_obs.json#L15-L27)

Would be included in the `CMOR_input.json` file, whereas all registered content would be used by default, and if there was an additional definition of the variable with the same name in the `CMOR_input.json` then this would overwrite the `input4MIPs_CV.json` sourced entry.

Can you comment on this?

@taylor13 ping

**Comment by durack1** [comment]
@mauzey1 specfically I have attempted to remove the variables: `calendar, contact, dataset_category, further_info_url, grid_label, nominal_resolution, region` and `source` from the `CMOR_input.json` file and CMOR 3.6.0 complains, even though these are read from the `PCMDI-AMIP-1-2-0` source_id registered in `input4MIPs_CV.json`

**Comment by durack1** [comment]
I vaguely remember the code dealt with this in a fairly clumsy way, but for a user it would seem that the information being fed to CMOR is somewhat arbitrary, so some variables can existing in the registered content (`input4MIPs_CV.json`) alone, whereas other variables MUST existing in the input (`CMOR_input.json`) even if these are duplicated

**Comment by durack1** [comment]
I've just stumbled across an interesting situation where I have removed my `frequency` from the `CMOR_input.json`, added this to the registered content (`input4MIPs_CV.json`) and now CMOR ignores the value of `"frequency":"fx",` in the table file that is being called to write data, this seems like an incorrect behavior - in the case of a variable defined in a table (in this case `input4MIPs_Ofx.json`) the priority should always be for the table value.

So in order, information should be overwritten from highest to lowest priority:
`input4MIPs_Ofx.json/table input  <-|  CMOR_input.json/user input <-| input4MIPs_CV.json/registered content`

**Comment by mauzey1** [comment]
@durack1 

So the CV file will be the default for input values unless set differently in the input file, but the variable's table entry will always precedence?

Should this behavior also apply to the CMIP6 CV and variable tables?

Where else is the `calendar` parameter defined aside `CMOR_input.json`?  I don't see it anywhere else.  Does it default to `gregorian`?

**Comment by durack1** [comment]
@mauzey1 yeah sorry, I haven't pushed all my latest versions to github, (have now done and links https://github.com/PCMDI/cmor/issues/628#issue-987134993 now point to the branch) - but the latest version on disk of the CV file does contain `calendar`. Regarding defaults, I presume that gregorian would be a reasonable default, but to be honest, considering CMOR is most often used to write model data, it may be something that we really don't want to assume and define default values

**Comment by durack1** [comment]
@mauzey1 if you wanted me to hand over a test env (including data, table files, input json etc), happy to provide you a directory(ies) to point to

**Comment by mauzey1** [comment]
@durack1 Yes, the test environment would be helpful to have.  Thank you.

**Comment by durack1** [comment]
@mauzey1 do you have access to detect, or which machine would you prefer? As we now have all our home co-located, this should be as easy as divulging a path and we're off - can you see /home/[me]? and /work/[me]

**Comment by mauzey1** [comment]
I don't have access to detect.  Could you put the env in a zip file and and email it to me?

**Comment by taylor13** [comment]
Regarding https://github.com/PCMDI/cmor/issues/628#issuecomment-912127523, I favor  *not* providing a default value for ``calendar'' for CMIP.  We would like to force the modelers to tell us what their calendar actually is (and many do not use ``gregorian''.  [I guess an error should be raised if they fail to supply ``calendar''.

**Comment by matthew-mizielinski** [comment]
> Regarding [#628 (comment)](https://github.com/PCMDI/cmor/issues/628#issuecomment-912127523), I favor _not_ providing a default value for `calendar` for CMIP. We would like to force the modelers to tell us what their calendar actually is (and many do not use `gregorian`. [I guess an error should be raised if they fail to supply `calendar`.

Absolutely -- I'm still getting queries from users who don't quite understand non-gregorian calendars when we have got this correct in our data. Providing a default seems a little risky to me.

**Comment by durack1** [comment]
@mauzey1 attached is a standalone demo, this should get us cooking with a demo that can elucidate the input data source priorities.

The input4MIPs-cmor-table symlink, will need to be cloned from https://github.com/PCMDI/input4MIPs-cmor-tables

Let me know if this is working for you, I've been testing against CMOR 3.6.0.

[210909_durack1_CMORTesting-issue628.zip](https://github.com/PCMDI/cmor/files/7145990/210909_durack1_CMORTesting-issue628.zip)

**Comment by durack1** [comment]
Depends on https://github.com/PCMDI/cmor/issues/629

**Comment by durack1** [comment]
Hi @mauzey1 let me know when you have time to start working on this, and I'll commit to pulling down a pre-release for testing

**Comment by mauzey1** [comment]
@durack1 I will try to get this feature ready for testing sometime this week.  I will keep you informed.

**Comment by durack1** [comment]
@mauzey1 perfect!

**Comment by durack1** [comment]
@mauzey1 I am starting to get close to having final data for writing, is there a demo/nightly 3.7.0 ready to roll yet?

**Comment by durack1** [comment]
@mauzey1 I am wondering whether we need to talk about this, as I am a little confused by the questions above.

The issue that I was trying to highlight is that the **source** (or the information source) of the information is what needs prioritizing, such that we load information in order, and overwrite existing variables as we go.

So for example, if `contact = bob@thebuilder.com` was found in the `input4MIPs_CV.json/registered content` file this would be loaded into memory. If the `CMOR_input.json/user input` file then contained the same variable with a different value `contact = bobby@thebuilder.com`, then a warning would be thrown and the existing instance of `contact` would be updated by the higher priority `user input`.

In the case that a variable was defined in the table `input4MIPs_Ofx.json/table input`, then this would overwrite the existing variable (in the example above `contact` (which is not included in the table definitions, but another variable, e.g. frequency would be for example).

So the intention, is to gather as much information from the registered content/CVs, so that a user can provide only limited information (not registered in the CVs) in the user input, BUT also importantly allow them to overwrite information at runtime.

Is this clear? I hope so

**Comment by durack1** [comment]
@mauzey1 regarding the specific `region` question, for input4MIPs we have no cases of multiple `region` values (see [here](https://github.com/PCMDI/input4MIPs-cmor-tables/blob/master/input4MIPs_source_id.json)), but this is not true for obs4MIPs for e.g. (see [here](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/master/obs4MIPs_source_id.json#L220-L223))

**Comment by mauzey1** [comment]
@durack1 

Okay.  So we initialize the attributes of a variable using the CV/registered content, but if the user input has that attribute we overwrite and throw a warning?  I think I have the part that sets the attributes with the CV/registered content already done.  I would just need to make something that throws the warning if the value is overwritten by the user input.  Is the list of attributes I posted that apply for this correct?

So the tables would overwrite user input?  Should that also throw a warning about overwriting?

**Comment by durack1** [comment]
@mauzey1 do we need to define a list or variables/attributes? The information will be slightly different depending on the project, so CMIP6/* vs input* vs obs4MIPs etc. If it is possible, a general `if variable exists, is overwritten, throw warning`

The tables should always be the definitive source of information, so if the table file defines it, then it overwrites all else regardless of the source (and throws a warning to alert the user they are doing bad things). At this point, it may be worth throwing an error and aborting, rather than a warning and continuing - @taylor13 @matthew-mizielinski any perspectives to share here?

**Comment by mauzey1** [comment]
@durack1

Okay.  I think what we could is go through the list of `required_global_attributes` to see what is required that is listed i the `source_id` in the CV.  This would eliminate the need to list the attributes since they are already listed in the CV.

Regarding the multiple `region` values in https://github.com/PCMDI/cmor/issues/628#issuecomment-960127709, would we store a list of regions in this attribute or just pick one?  Is there an example of multiple regions being stored in a file?  The rest of the attributes are easy since they are just one string but I wanted to know how to handle the list-of-strings case.

**Comment by durack1** [comment]
@mauzey1 the above 1 (go through the `required_global_attributes`) is exactly the right approach, however, if there are additional entries that are "optional" in the `source_id` then these should be loaded two, and written to the file in the case that no other instances exist in the higher priority input sources (input json, and table json).

Regarding `region` in obs4MIPs we do have the regions identified in [PCMDI/obs4MIPs-cmor-tables/obs4MIPs_region.json](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/master/obs4MIPs_region.json), and if this was included in other table/CV repos then the same listing of valid entries would be necessary (and then propagate into the `**PROJECT**_CV.json` file

**Comment by mauzey1** [comment]
In that case, should we just go through entire list of attributes for a `source_id` entry to include both required and optional attributes?  I'm guessing we can skip the `source_variables` attribute since it's just there to list which variables are part of the source, right?

I don't think that really answers my question.  I was asking if we should write a list of regions into a file's `region` global attribute if we encounter multiple regions listed for a `region` attribute in the `source_id` entry of a CV file.

**Comment by durack1** [comment]
> In that case, should we just go through entire list of attributes for a `source_id` entry to include both required and optional attributes? I'm guessing we can skip the `source_variables` attribute since it's just there to list which variables are part of the source, right?

Exactly, I would recommend that we include all the registered/CV entries, including `source_variables` because they are in the CV. As an aside, is there a way to overwrite/remove such a variable in the `cmor_input.json` - so a way to clobber the CV/registered values at run time? That would be useful. Would setting these to `None` be a way to do it?

> I don't think that really answers my question. I was asking if we should write a list of regions into a file's `region` global attribute if we encounter multiple regions listed for a `region` attribute in the `source_id` entry of a CV file.

That is a good question. In this case, I would write whatever is provided, so in the case that the CV has region specified, but the `cmor_input.json` doesn't, then I would write the CV attribute. For most users, they could then control this attribute by overwriting the value in the `cmor_input.json` file, clobbering the CV sourced value

Does that answer the remaining questions?

**Comment by taylor13** [comment]
If I understand it, you plan to write as  global attributes everything that is in the source_id section of the [input4MIPs_CV.json file](https://github.com/PCMDI/input4MIPs-cmor-tables/blob/issue87_durack1_PCMDI-AMIP-1-2-0AndPy3Update/Tables/input4MIPs_CV.json#L974-L1007).  That might be o.k. for input4MIPs, but for CMIP6, we don't want CMOR to write most of the information in the source_id section of the [cmip6_CV.json file](https://github.com/PCMDI/cmip6-cmor-tables/blob/f1f9be8afb56a7998925b1623d269b465d47daed/Tables/CMIP6_CV.json).  Isn't that right @durack1?

**Comment by durack1** [comment]
@taylor13 exactly. For CMIP6, the information that is contained in the `PCMDI/cmip6-cmor-tables/CMIP6_CV.json/source_id` entry is curated, so that it is a subset of the info in the CMIP6_CVs, so for e.g. `ACCESS-CM2`:

https://github.com/PCMDI/cmip6-cmor-tables/blob/f35e08541ed75201e0063ddec58daa361b9dfa5e/Tables/CMIP6_CV.json#L140-L158

vs the original/registered info (in the [`WCRP-CMIP/CMIP6_CVs/CMIP6_source_id.json`](https://github.com/WCRP-CMIP/CMIP6_CVs/blob/master/CMIP6_source_id.json)):

https://github.com/WCRP-CMIP/CMIP6_CVs/blob/7fb62c418c8323bbda6cc327bbd1523e8fd810f5/CMIP6_source_id.json#L52-L106

**Comment by taylor13** [comment]
Got it.  But I think we should then filter out "cohort".  We don't want CMOR to write that as a global attribute because its status will change over time (e.g., from "registered" to something else (can't remember) once all the DECK experiments have been completed.

**Comment by durack1** [comment]
> Got it. But I think we should then filter out "cohort". We don't want CMOR to write that as a global attribute because its status will change over time (e.g., from "registered" to something else (can't remember) once all the DECK experiments have been completed.

Agreed, I wonder if we need to create a new `excluded_global_attributes` which means we can control such things in the `CMIP6_CV.json` file, we do have the [`required_global_attributes`](https://github.com/PCMDI/cmip6-cmor-tables/blob/f35e08541ed75201e0063ddec58daa361b9dfa5e/Tables/CMIP6_CV.json#L3-L34)

To be honest, we may just want to remove `cohort` from the entry, but having said that, it's already part of most of the files that have used the default CMOR3 configuration - too late to change?

**Comment by taylor13** [comment]
Is it really true that CMOR currently writes all global attributes defined in CMIP6_CV.json?  I thought it only wrote the required ones.  In any case I think we definitely do *not* want to include "cohort" in future files written by CMOR.  The fact that current files might have flaws doesn't mean we don't want to eliminate those in future files.

**Comment by durack1** [comment]
I'll take a look at the `ACCESS-*` output later today, I'm fairly sure they're using the CMOR3 pipeline with minimum edits, my guess is yes, unless an exclusion was made in CMOR many moons ago

**Comment by durack1** [comment]
Looks like you did manage to get it excluded, at least in this example:
```
(base) -bash-4.2$ ncdump -h ~CMIP6/CMIP/CSIRO/ACCESS-ESM1-5/historical/r1i1p1f1/Omon/tos/gn/v20191115/tos_Omon_ACCESS-ESM1-5_historical_r1i1p1f1_gn_185001-201412.nc 
netcdf tos_Omon_ACCESS-ESM1-5_historical_r1i1p1f1_gn_185001-201412 {
dimensions:
	time = UNLIMITED ; // (1980 currently)
	j = 300 ;
	i = 360 ;
	bnds = 2 ;
	vertices = 4 ;
...

// global attributes:
		:Conventions = "CF-1.7 CMIP-6.2" ;
		:activity_id = "CMIP" ;
		:branch_method = "standard" ;
		:branch_time_in_child = 0. ;
		:branch_time_in_parent = 21915. ;
		:creation_date = "2019-11-15T15:21:42Z" ;
		:data_specs_version = "01.00.30" ;
		:experiment = "all-forcing simulation of the recent past" ;
		:experiment_id = "historical" ;
		:external_variables = "areacello" ;
		:forcing_index = 1 ;
		:frequency = "mon" ;
		:further_info_url = "https://furtherinfo.es-doc.org/CMIP6.CSIRO.ACCESS-ESM1-5.historical.none.r1i1p1f1" ;
		:grid = "native atmosphere N96 grid (145x192 latxlon)" ;
		:grid_label = "gn" ;
		:history = "2019-11-15T15:21:42Z ; CMOR rewrote data to be consistent with CMIP6, CF-1.7 CMIP-6.2 and CF standards." ;
		:initialization_index = 1 ;
		:institution = "Commonwealth Scientific and Industrial Research Organisation, Aspendale, Victoria 3195, Australia" ;
		:institution_id = "CSIRO" ;
		:mip_era = "CMIP6" ;
		:nominal_resolution = "250 km" ;
		:notes = "Exp: ESM-historical; Local ID: HI-05; Variable: tos ([\'sst\'])" ;
		:parent_activity_id = "CMIP" ;
		:parent_experiment_id = "piControl" ;
		:parent_mip_era = "CMIP6" ;
		:parent_source_id = "ACCESS-ESM1-5" ;
		:parent_time_units = "days since 0101-1-1" ;
		:parent_variant_label = "r1i1p1f1" ;
		:physics_index = 1 ;
		:product = "model-output" ;
		:realization_index = 1 ;
		:realm = "ocean" ;
		:run_variant = "forcing: GHG, Oz, SA, Sl, Vl, BC, OC, (GHG = CO2, N2O, CH4, CFC11, CFC12, CFC113, HCFC22, HFC125, HFC134a)" ;
		:source = "ACCESS-ESM1.5 (2019): \n",
			"aerosol: CLASSIC (v1.0)\n",
			"atmos: HadGAM2 (r1.1, N96; 192 x 145 longitude/latitude; 38 levels; top level 39255 m)\n",
			"atmosChem: none\n",
			"land: CABLE2.4\n",
			"landIce: none\n",
			"ocean: ACCESS-OM2 (MOM5, tripolar primarily 1deg; 360 x 300 longitude/latitude; 50 levels; top grid cell 0-10 m)\n",
			"ocnBgchem: WOMBAT (same grid as ocean)\n",
			"seaIce: CICE4.1 (same grid as ocean)" ;
		:source_id = "ACCESS-ESM1-5" ;
		:source_type = "AOGCM" ;
		:sub_experiment = "none" ;
		:sub_experiment_id = "none" ;
		:table_id = "Omon" ;
		:table_info = "Creation Date:(30 April 2019) MD5:e14f55f257cceafb2523e41244962371" ;
		:title = "ACCESS-ESM1-5 output prepared for CMIP6" ;
		:variable_id = "tos" ;
		:variant_label = "r1i1p1f1" ;
		:version = "v20191115" ;
		:cmor_version = "3.4.0" ;
		:_NCProperties = "version=2,netcdf=4.6.2,hdf5=1.10.5" ;
		:tracking_id = "hdl:21.14100/02850fcc-be64-40de-b7ca-9b8aa6e688a0" ;
		:license = "CMIP6 model data produced by CSIRO is licensed under a Creative Commons Attribution-ShareAlike 4.0 International License (https://creativecommons.org/licenses/).  Consult https://pcmdi.llnl.gov/CMIP6/TermsOfUse for terms of use governing CMIP6 output, including citation requirements and proper acknowledgment.  Further information about this data, including some limitations, can be found via the further_info_url (recorded as a global attribute in this file).  The data producers and data providers make no warranty, either express or implied, including, but not limited to, warranties of merchantability and fitness for a particular purpose. All liabilities arising from the supply of the information (including any liability arising in negligence) are excluded to the fullest extent permitted by law." ;
}
```

We'll have to figure out where the exclusion is hard-coded, might be something else to clean up.

**Comment by mauzey1** [comment]
@durack1 

So far, I have made it so that the attributes in the `source_id` entry of the CV are used in the global attributes of the dataset unless the attributes are defined in the user input file.  However, this lead to some issues with the attributes in the entries that were lists of strings such as `region` and `source_variables`.  A part of CMOR will check if the required attributes' values are listed in the CV like the `region` values in [input4MIPs_CV.json](https://github.com/PCMDI/input4MIPs-cmor-tables/blob/41a6c98858a3c3944fe91139b3bac27e2c336179/Tables/input4MIPs_CV.json#L149-L218).

When I tried to create a string value that just all the region values listed in the source_id entry, CMOR treated it like as an error.  I could just use the first value since the `region` attribute in input4MIPs variables just have one listed.  However, there could be multiple values listed in other CVs.  The `source_variables` attribute would be ignored since it is not considered a required attribute in the CV.  I could make a special case for attributes like `region` where it checks if the value is either listed in the list, or is a list as defined in the `source_id` entry.  However, that might run into an issue if the string exceeds the current string maximum length of 1024.

As for using the variable table values to overwrite the user input, would that be just for attributes listed in the table header or variable entry that are also listed in the `required_global_attributes` of the CV?

**Comment by durack1** [comment]
@mauzey1 this sounds like progress, I am doubtful that `region` would exceed 1024 characters, but you never know!

Regarding the table values, we don't want any of this information to be overwritten, so in the case of `Omon | tos`:
```
{
    "Header":{
        "Conventions":"CF-1.7 CMIP-6.2",
        "approx_interval":"30.00000",
        "cmor_version":"3.5",
        "data_specs_version":"01.00.32",
        "generic_levels":"olevel olevhalf",
        "int_missing_value":"-999",
        "mip_era":"CMIP6",
        "missing_value":"1e20",
        "product":"input4MIPs",
        "realm":"ocean",
        "table_date":"14 September 2020",
        "table_id":"Table input4MIPs_Omon"
    },
    "variable_entry":{
        "tos":{
            "cell_measures":"area: areacello",
            "cell_methods":"time: mean",
            "comment":"",
            "dimensions":"longitude latitude time",
            "frequency":"mon",
            "long_name":"Sea Surface Temperature",
            "modeling_realm":"ocean",
            "ok_max_mean_abs":"",
            "ok_min_mean_abs":"",
            "out_name":"tos",
            "positive":"",
            "standard_name":"sea_surface_temperature",
            "type":"real",
            "units":"degC",
            "valid_max":"",
            "valid_min":""
        },
```
We want to ensure that all data using these tables conforms to this information verbatim, and no user-provided input can overwrite any of these details, throwing an error (and exiting) if this is attempted

**Comment by mauzey1** [comment]
@durack1 Do you have any examples of the attribute values from the variable table being changed by the user input or via the API?  I tried to change the `Conventions` attribute in the user input file for the input4MIPs example program but it doesn't change the global attribute value in the NetCDF file.

cmor_input.json
```
    "institution_id":               "PCMDI",
    "source_id":                    "PCMDI-AMIP-1-2-0",
    "Conventions":                  "CF-XXXXXXXXXX",
```
osbcs_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-2-0_gn_187001-187512.nc
```
// global attributes:
                :Conventions = "CF-1.7 CMIP-6.2" ;
```

**Comment by durack1** [comment]
@mauzey1 apologies for the tardy reply.

The code in https://github.com/PCMDI/amipbcs/blob/issue23_durack1_cleanupCMORinput/sanitize.py#L442 was how I was testing, and the last time I tested it overwrote the table loaded value (I think)

**Comment by mauzey1** [comment]
@durack1

I'm still having issues with changing the attributes with either the user input file or the API.  I tried inserting the line `cmor.set_cur_dataset_attribute('frequency', 'fx')` in the sanitizeTest.py file you provided previously.
```
# Force local file attribute as history
cmor.set_cur_dataset_attribute('history', history)
cmor.set_cur_dataset_attribute('frequency', 'fx')
table = 'input4MIPs_Omon.json'
cmor.load_table(table)
```
It doesn't seem to change anything since the `frequency` global attribute is still set to `mon` in the output file.

**Comment by durack1** [comment]
@mauzey1 is the order of the inputs set in the code somewhere? That might be a way to understand what is defined and redefined/overwritten etc

**Comment by mauzey1** [comment]
@durack1 So I think I have figured out where and how attributes get set/overwritten.  Although you can set attributes in the user input or inside the code with the CMOR API, the function `cmor_CV_checkGblAttributes`, that is called in `cmor_setGblAttr` from `cmor_write`, will set the required global attributes to those found in the `source_id` entry of the CV.

When `cmor_CV_checkGblAttributes` is called, it will loop through all attributes listed in `required_global_attributes` of the CV file to check if the attributes are set.  If they are not set, then it will throw an error.  If set, then it will run the function `cmor_CV_ValidateAttribute`.  `cmor_CV_ValidateAttribute` will either check if an attributes value is in a list in the CV, or populate the other attributes if the checked attribute is the key to another set of attributes in the CV file i.e. `source_id`.  However, it won't set values for attributes that are list such as `region`.

I think the process for adding registered content should go as follows.  At the beginning when the CV file is loaded, CMOR will loop through all of the attributes listed in the `source_id` entry.  Afterwards, it will allow for the attributes to be modified by either the user input file or CMOR API calls.  When CMOR reaches the `cmor_CV_ValidateAttribute` function, it will check whether there is a value already set for an attribute before attempting to set it.  If the value differs from that of the CV's registered content, then a warning will be thrown indicating that the user has overwritten the registered content of the attribute.

As for changing the attributes that are listed in the header of the variable's table file, they seem to be unable to change no matter where I call `set_cur_dataset_attribute` in the sanitizeTest.py program.  I tried to change the value of `realm` in a few places in the code but it still stays the same value as listed in the table header when written to file.  It already seems pretty hard for a user to alter the table attributes via CMOR.

**Comment by durack1** [comment]
@mauzey1 apologies for the tardy reply. What you outline above sounds perfect, I'll ping @taylor13 here just for a sanity check.

I believe my test was attempting to overwrite the `frequency` attribute after loading the table file, and before writing the data, but you've not been able to recreate my experience, so I should double check once we have a nightly version to test

**Comment by taylor13** [comment]
@mauzey1 @durack1 :  Re
```
I think the process for adding registered content should go as follows. At the beginning when 
the CV file is loaded, CMOR will loop through all of the attributes listed in the source_id entry. 
Afterwards, it will allow for the attributes to be modified by either the user input file or CMOR 
API calls. When CMOR reaches the cmor_CV_ValidateAttribute function, it will check whether 
there is a value already set for an attribute before attempting to set it. If the value differs from 
that of the CV's registered content, then a warning will be thrown indicating that the user has 
overwritten the registered content of the attribute.
```
I think this is about right, but not quite. Here is how I see it:
1) I think CMOR might only care about a subset of the information under each source_id entry in the CV file.   (For example, CMOR might not care about "cohort".)
2) For most, if not all, attributes in the source_id CV, the user should not be permitted to make *any* changes when running CMOR.  The CV is supposed to provide the *reference* information for the source.  If something is incorrect in the CV, then it should be changed in the source_id CV itself prior to running CMOR.
3)  For most, if not all, attributes in the CMOR tables, the user should not be able to make *any* changes when running CMOR.  Again these tables represent the definitive source of information pertaining to a variable.
4) If a user's cmor input file or a CMOR API call attempts to change a value obtained from the source_id CV or from a cmor table, a *fatal* error should be raised.

Do either of you agree with me?

I also have a somewhat unrelated question:  Does CMOR record activity_id based on the experiment_id  CV table, which contains the activities associated with an experiment?  If so, I don't think the user should be able to override the value in the CMIP6_input.json input table to cmor.

**Comment by taylor13** [comment]
point 3 above is not quite right.  The global attributes and *most* of the variable attributes set by the CMOR tables should not be overridden.  There is some variable-specific information that we should allow to be overridden (e.g., the value of a singleton dimension, like the height where a surface variable is measured).

**Comment by mauzey1** [comment]
@durack1 @taylor13 

Okay.  So instead of just displaying a warning for having a value for an attribute differing from that found in the `source_id` entry of the CV file, we should be throwing an error?

For the values in the `source_id` entry that are in arrays (values in brackets such as `region` or `source_variables`), do we just use those to validate the input?  Or do we attempt to set attributes with the values in those arrays, either using the first value in the array or the entire array as a list?

**Comment by wachsylon** [comment]
One additional comment here for `further_info_url` (if not already treated). We will use CMOR3 for other projects and do not want to create `further_info_url`s. However, I think that it is not possible to remove it at all as it is created and checked on the same level as e.g. `output_file_template` [here](https://github.com/PCMDI/cmor/blob/fe8ade4fca4b9b31a8aab5c1c344798ce5cb3722/Src/cmor.c#L1323) with [this](https://github.com/PCMDI/cmor/blob/fe8ade4fca4b9b31a8aab5c1c344798ce5cb3722/include/cmor.h#L266). I do not understand why this attribute is of such importance that it must be created when CMOR is applied so I would prefer to remove that from both the header and source.

**Comment by mauzey1** [comment]
So far, I have made CMOR display warnings when it detects that a global attribute deviates from what is listed in the `source_id` entry of the CV file.  This is assuming that the value in the entry is a single string and not an array.  Below is what happens when the attribute `product` is set to `derived` in the user input file of the input4MIPs test.
```
varName: tosbcs units: degC axis_ids: [0, 1, 2]

C Traceback:
In function: _CV_ValidateAttribute
! called from: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Warning: The registered CV attribute "product" as defined as "observations" will be replaced with 
! "derived" as defined in your user input file
! 
!
!!!!!!!!!!!!!!!!!!!!!!!!!

! ------
! CMOR is now closed.
! ------
! During execution we encountered:
!   1 Warning(s)
!   0 Error(s)
! ------
! Please review them.
! ------
! 
```
This results in a NetCDF file where the global attribute `product` is set to `derived`.  If I remove `product` from the user input file, then it would default to the value `observations`.  However, the value listed in the table's header for `product` is `input4MIPs`.
https://github.com/PCMDI/input4MIPs-cmor-tables/blob/17163f642582a4794551703158c03e9f2d478472/Tables/input4MIPs_Omon.json#L11
Wouldn't this cause an issue with the part where we throw an error if the attribute's value deviates from that listed in the table?

**Comment by taylor13** [comment]
I'm not that familiar with the input4MIPs CVs, but I think if the CV says this source provides "observations", then the user should *not* be able to change that to "derived".  Instead the user should request that the source reference documentation be changed to indicate that the product is "derived", not "observations".

**Comment by mauzey1** [comment]
Okay, so the warning should be elevated to an error indicating the correct value that an attribute should be?

**Comment by durack1** [comment]
This maybe an inconsistency that requires a cleanup in the input4mips-CMOR-tables. I've not thought much about this, but will review. Not all forcing datasets are "observations" either

**Comment by taylor13** [comment]
yes, let's give this some thought before making changes.

**Comment by mauzey1** [comment]
@durack1 

So far, I have a [branch](https://github.com/PCMDI/cmor/tree/628_cleaning_cmor_input) in the repo that you can test.  It currently fulfills the use of the attributes in the `source_id` entry of the CV to fill global attributes.  It doesn't perform the part where it rejects values of attributes defined in the variable table if they don't match.

I would prefer that changes get reviewed in braches before I push them to the nightly build.

**Comment by mauzey1** [comment]
@durack1 

I've created a new `testing` label for builds that contain experimental features that need testing before being added to the `nightly` build.  To create a conda environment for testing it with Python, run the following.
```
conda create -n cmor_testing -c pcmdi/label/testing -c conda-forge -c cdat/label/nightly -c cdat cmor cdms2 testsrunner
```
This build has the changes I made in the [branch](https://github.com/PCMDI/cmor/tree/628_cleaning_cmor_input) I mentioned previously.

**Comment by durack1** [comment]
@mauzey1 perfect!

I assume the correct package is labelled (for py3.9) `osx-64/cmor-3.6.1.2022.01.06.03.14.g3926ce5-py39h479cae4_0.tar.bz2`? - great and that also lines up with the `testing` branch at https://github.com/PCMDI/cmor/tree/testing, so this should be something that we can easily keep trace of over new conda builds

**Comment by mauzey1** [comment]
@durack1 Were you able to test this build of CMOR?

**Comment by durack1** [comment]
@mauzey1 that's on the to-do list for this week, I was keen to get this data written and published pronto - will circle back when I have feedback for you

**Comment by durack1** [comment]
@mauzey1 just confirming, I have just created a new env with
```
+ cmor 3.6.1.2022.01.06.04.48.g3926ce5 py310h88780f8_0 pcmdi/label/testing/osx-64 956 KB
```
I assume I'm playing with the right version?

**Comment by mauzey1** [comment]
@durack1 Yes, that is the right version.

**Comment by durack1** [comment]
@mauzey1, I am just working through my last test, which led to this issue. In the tables, we have `mip_era` defined, so for e.g. 
`master`: ["mip_era":"CMIP6",](https://github.com/PCMDI/input4MIPs-cmor-tables/blob/master/Tables/input4MIPs_Omon.json#L9)
`issue87 branch`: ["mip_era":"CMIP6Plus",](https://github.com/PCMDI/input4MIPs-cmor-tables/blob/issue87_durack1_PCMDI-AMIP-1-2-0AndPy3Update/Tables/input4MIPs_Omon.json#L9)

Is it possible to remove `mip_era` from the table header? And provide this in the user input, or better, in the registered CVs? Or alternatively, could `"mip_era":"CMIP6 CMIP6Plus",` work? (indicating that either `mip_era` was an accepted value?

**Comment by mauzey1** [comment]
@durack1 

I did an experiment where I changed the `mip_era` for the table and user input to `CMIP6` and `CMIP5` respectively.  If I run sanitizeTest.py with this configuration, I get the following message.
```
C Traceback:
In function: _CV_ValidateAttribute
! called from: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Warning: The registered CV attribute "mip_era" as defined as "CMIP6Plus" will be replaced with 
! "CMIP6" as defined in your user input file
! 
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```
This will create a file with `CMIP6` as its mip era.

If I were to remove the `"mip_era":"CMIP6"` line from the header of input4MIPs_Omon.json, then I get the following message.
```
C Traceback:
In function: _CV_ValidateAttribute
! called from: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Warning: The registered CV attribute "mip_era" as defined as "CMIP6Plus" will be replaced with 
! "CMIP5" as defined in your user input file
! 
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```
This will create a file with `CMIP5` as its mip era.

Removing `"mip_era": "CMIP5"` from the user input will make CMOR generate a file with `CMIP6Plus` as its mip era as defined in input4MIPs_CV.json.

It seems like CMOR is setting `mip_era` where the registered CV value is used by default but is overridden by the value defined in user input, which in turn is overridden by the table's value.  Removing the table's `mip_era` would mean the CV's value would be used if `mip_era` is not defined in the user input.

**Comment by mauzey1** [comment]
> Or alternatively, could `"mip_era":"CMIP6 CMIP6Plus",` work? (indicating that either `mip_era` was an accepted value?
@durack1 You mean having having a list of valid values for `mip_era` for a `source_id` entry in the CV, or in a variable's table?

**Comment by durack1** [comment]
@mauzey1 yes sorry to be more clear. I was hoping that we could generate tables that could be used across CMIP6 and CMIP6Plus, so not requiring to have two sets of identical tables for each mip_era. I did try adding both entries (space delimited) and that didn't seem to work

**Comment by mauzey1** [comment]
@durack1 So would we put `"mip_era":"CMIP6 CMIP6Plus"` in the header of a table (Ex. input4MIPs_Omon.json) to indicate that it could be used to create files with either mip era?  Would that mean that the mip era values listed in the user input and CV are what get used to create the file, and they must be one of the mip era values listed in the table's header?

**Comment by durack1** [comment]
@mauzey1 exactly. We need to think a little about this, as it's the first time (I think) that we'd try to reuse tables to cover two mip_era entries. @taylor13 and I will have to put our heads together and think this through a little.

@matthew-mizielinski pinging you on this thread too

**Comment by taylor13** [comment]
My current view is that "mip_era" is not a particularly good descriptor for distinguishing one dataset from another.  For experiments like abrupt4xCO2, the experiment design is not expected to change significantly over time, so classification by "era" seems pointless.  (Note that at any point in time a very old model or a recently developed model might run this experiment, so era isn't a good indication of the "model's newness" either.  

A much more important and fundamental distinction between abrupt4xCO2 model output generated for CMIP5 and model output generated for CMIP6 is that the DRS is different.  I therefore suggest that in place of labeling output with "mip_era", we explicitly interpret "mip_era" as identifying the DRS that a dataset is compliant with.  (I would also suggest that in the future, we rename this descriptor "DRS_rules" (which would be set to "mip_era" (i.e., "CMIP6") in the current model archive.  (For input4MIPs, we would define "DRS_rules"="input4MIPs6", or something else appropriate and different from "CMIP6");  for "obs4MIPs", "DRS_rules"="ODS2.1"; and for CMIP5, DRS_rules="CMIP5").

**Comment by durack1** [comment]
@mauzey1 apologies for taking my time on this, I am now starting to work on generating this new data so am following up where I left off and will use this issue to drop down a couple of notes before trying to new `testing` pre-release versions.

When I remove `"source"` from the `cmor_input.json` file I get:
```

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: Your Control Vocabulary file specifies one or more
! required attributes.  The following
! attribute was not properly set.
! 
! Please set attribute: "source" in your input file.
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```
Even though this is defined in the registered content, identified by the `source_id` label in the `cmor_input.json` file

**Comment by mauzey1** [comment]
@durack1 That error should not be appearing the `testing` build.

**Comment by durack1** [comment]
@mauzey1 I was hoping this was the case, that's next on the to-do list, the v1.1.7 data (rather than a reformatted v1.2.0) was published today, thanks @sashakames

**Comment by durack1** [comment]
@mauzey1 first up apologies for taking this long to get back to this.

Ok, I have now pulled down `CMOR 3.7.0 pre-release/testing` (cmor 3.6.1.2022.02.25.17.07.g963fd11 py310h8466d85_0 pcmdi/label/testing/linux-64) and have updated my input deck to reflect the changes that I had proposed, which I will describe again below

1. I have updated the [PCMDI/input4mips-cmor-tables/input4MIPs_source_id.json](https://github.com/PCMDI/input4MIPs-cmor-tables/blob/2ff3f0a45eb5a2e8e46eee4252261d74aeb1add3/input4MIPs_source_id.json#L766-L800) so that all the required global attributes are in the `PCMDI-AMIP-1-2-0` registration
2. I have removed these entries from the `user_input.json` file (in an attempt to depend wholly on the registered content), which can be found at [PCMDI/amipbcs/CMOR/drive_input4MIPs_bcs.json](https://github.com/PCMDI/amipbcs/blob/issue25_durack1_CMOR3p7InputTweakTest/CMOR/drive_input4MIPs_bcs.json), the diff can be seen on the PR [here](https://github.com/PCMDI/amipbcs/pull/26/files)

This now gives an error, the `region` variable is defined in the `PCMDI-AMIP-1-2-0` source_id registration, not the user_input.json, so it should be available for CMOR to use, but bombs:

```python
C Traceback:
! In function: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: Your Control Vocabulary file specifies one or more
! required attributes.  The following
! attribute was not properly set.
! 
! Please set attribute: "region" in your input file.
!
!!!!!!!!!!!!!!!!!!!!!!!!!

C Traceback:
! In function: cmor_get_cur_dataset_attribute
! called from: _CV_ValidateAttribute
! called from: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: Dataset: current dataset does not have attribute : region
!
!!!!!!!!!!!!!!!!!!!!!!!!!

C Traceback:
In function: _CV_ValidateAttribute
! called from: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Warning: The registered CV attribute "mip_era" as defined as "CMIP6Plus" will be replaced with 
! "CMIP6" as defined in your user input file
!
!!!!!!!!!!!!!!!!!!!!!!!!!

C Traceback:
! In function: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: Please fix required attributes mentioned in
! the warnings/error above and rerun. (aborting!)
!
!!!!!!!!!!!!!!!!!!!!!!!!!

C Traceback:
! In function: 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: Cannot continue until you fix the errors listed above: -2
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```
I do note that this variable is defined as a json `list` type, not a `string` as such registrations can have more than one entry


A separate point, the error `"Warning: The registered CV attribute "mip_era" as defined as "CMIP6Plus" will be replaced with "CMIP6" as defined in your user input file"` is a little off, as this information is not provided in the `user input file` but is rather included in both the registered content/source_id input, plus the entry in the `input4MIPs_SImon/Omon.json` files in the `"mip_era":"CMIP6",` header entry

**Comment by mauzey1** [comment]
@durack1 

I have fixed the warning message for values from the tables.  I have made a change that will detect if the value of an attribute comes from a table rather than the user input.

```
C Traceback:
In function: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Warning: The registered CV attribute "mip_era" as defined as "CMIP6Plus" will be replaced with 
! "CMIP6" as defined in the table input4MIPs_Omon
! 
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```

I will push the change to the testing branch now.

**Comment by durack1** [comment]
@mauzey1 great, I am setup to update the conda env, so will pull down the latest when it lands - I presume this will be today?

**Comment by mauzey1** [comment]
@durack1 I would expect the build to be available in about 1-2 hours from now.  

I haven't addressed the issue with value list attributes like `region`, yet.  How would you like CMOR to handle these kinds of attributes?  Would having a comma-separated list of values in a string be a suitable value for this attribute?  Maybe just using the first attribute that appears on the list?  Are there examples in datasets where multiple string values are held in a single variable?  Having such list of values for an attribute would mean that PrePARE will need to properly identify such values.

**Comment by durack1** [comment]
@mauzey1 yes, I think a comma-separated string would be perfect. We have a couple of obs4MIPs datasets (see [here](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/master/obs4MIPs_source_id.json#L236-L240) that have multiple entries using the same list type, so managing that across all inputs would be useful (if it's not too hard - happy to iterate over this)

**Comment by durack1** [comment]
@sashakames just wondering how the ESGF publisher would deal with a comma-separated global attribute, are there any publisher-specific formats we need to ensure?

**Comment by mauzey1** [comment]
Since multiple values seem to be rare for these kind of attributes, maybe we should consider just using the only value present when it is just one in the list.  For multiple values in a list, we could give an error that warns the user that they need to choose one.

An example in the CMIP6 tables is `institution_id`.  This attribute in the registered data is in bracketed list form but most of them only have one value listed.  However, there are several entries that have multiple values like [this](https://github.com/PCMDI/cmip6-cmor-tables/blob/196f265c1c2f92db15da9d6f5d21e15010b281b0/Tables/CMIP6_CV.json#L510-L513).  If the institution id is not set, then CMOR could have an error message like the following:

```
!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: The registered CV attribute "institution_id" has multiple values defined in CMIP6_CV.json.
! Please use one of the following values for this attribute.
!            * UCI
!            * NCAR
! 
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```

The message could also simply ask the user to refer to the CV file for the valid values.

The attributes `cohort` and `activity_participation` are also list with `activity_participation` having many entries listed, but these attributes are not listed in [`required_global_attributes`](https://github.com/PCMDI/cmip6-cmor-tables/blob/196f265c1c2f92db15da9d6f5d21e15010b281b0/Tables/CMIP6_CV.json#L3-L34) so CMOR doesn't consider them.

**Comment by durack1** [comment]
@mauzey1 exactly, thanks for bringing up these examples. So for the CMIP6 project, we'd need a user to provide the following (in the case that multiple/list values are found in the registered content):
- `activity_participation`
- `institution_id`
- `region`

For input4MIPs the only case where [multiple registered values exist are](https://github.com/PCMDI/input4MIPs-cmor-tables/blob/master/input4MIPs_source_id.json):
- `region`
- `source_variables`

For obs4MIPs [same](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/master/obs4MIPs_source_id.json):

**Comment by sashakames** [comment]
@durack1 We are using a space as the delimiter now.  For old publisher users, the config can be updated, but we have been using an official config and this would mean two versions potentially.  As the new publisher is in beta, I have some flexibility to change how we configure the project so users would be able to gradually migrate to such a change if you decide to go ahead.

**Comment by sashakames** [comment]
Some more background: for input4MIPs we were going to use commas. Instead we just get the values out of the .json file so it doesn't read the global attributes to generate additional facet metadata (beyond those in the DRS)

**Comment by mauzey1** [comment]
@sashakames So there can be cases where a dataset will have attributes with multiple values?  Something like `institution_id: UCI NCAR` is possible?

**Comment by sashakames** [comment]
@mauzey1 Yes indeed.  I think `activity_id` and `realm` were the CMIP6 examples (not institution_id though to my knowledge)  Sorry I don't have an example but there are some out there.

**Comment by durack1** [comment]
@sashakames you're right, those two as well. The example of the activity_id is already caught in the source_id file, but realm, absolutely, there are a couple of limited example that noted both seaIce/ocean realms, and maybe a couple more - I could dig them up if useful @mauzey1

**Comment by durack1** [comment]
@matthew-mizielinski thanks for clarifying my post. @sashakames this comes back to the registered info (CMIP6_CVs), and then the info required by CMOR to write any one dataset.

For the registered info, we will need to allow a single `source_id` to have multiple `institution_id` entries. As is the case of  [`UKESM1-0-LL`](https://github.com/WCRP-CMIP/CMIP6_CVs/blob/master/CMIP6_source_id.json#L7345-L7369) this single `source_id` is being used by 4 registered `institution_id`'s, and each of these contributed datasets should use the appropriate `institution_id`, `contact` etc to ensure that the right folks are being contacted if data queries arise.

Thinking a little more about this, maybe it would be a good idea to check the [`output_path_template`](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_input_example.json#L72) field that is required by CMOR, and ensure that any output/DRS or filename components only have a single value. If multiple values are registered in the CMIP6_CVs, then we will stop throwing an error that the `CMOR_user_input.json` file requires a single value to be defined

**Comment by mauzey1** [comment]
@durack1 If an attribute has multiple values separated by a space, then CMOR will only use the first value in that list for the output path.
https://github.com/PCMDI/cmor/blob/90bde9eae22c893cf2c010c716ba5934d8d9005e/Src/cmor.c#L5561-L5572

**Comment by mauzey1** [comment]
Following up on this issue, I have made changes to CMOR that would allow it to populate attributes with multiple values if they are in a list in the CV.  However, I have tried running PrePARE on files where this change is applied.  For example, a file with `institution_id = "MOHC NERC NIMS-KMA NIWA"` gives the following output.
```
C Traceback:
! In function: _CV_setInstitution
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: The institution_id, "MOHC NERC NIMS-KMA NIWA",  found in your 
! input file () could not be found in 
! your Controlled Vocabulary file. (Tables/CMIP6_CV.json) 
! 
! Please correct your input file or contact "cmor@listserv.llnl.gov" to register
! a new institution_id.  
! 
! See "http://cmor.llnl.gov/mydoc_cmor3_CV/" for further information about
! the "institution_id" and "institution" global attributes.  
!
!!!!!!!!!!!!!!!!!!!!!!!!!

└──> :: CV FAIL    :: /Users/mauzey1/Desktop/github/cmor/fDeforestToProduct_Emon_UKESM1-0-LL_1pctCO2_r2i1p1f2_gn_195001-199912.nc

Number of files scanned: 1
Number of file with error(s): 1
```
Attributes like `source_type` have been programmed to handle having multiple values listed, independent of order listed.  Do we want to apply similar behavior to attributes such as `institution_id`?

**Comment by durack1** [comment]
Thanks @mauzey1!

Following on from what @taylor13 was noting, we have 4 classes of attributes

1. an attribute assigned a single value, unchecked
For example, mip_era = "CMIP6"
2. an attribute assigned a single value from a CV, checked
For example, source_id = "UKESM1-0-LL", where only registered models are allowed.
3. an attribute assigned a list of values, specified in a registry/CV/CMOR inputs, table header etc (nb. sometimes the "list" is a single item), checked
For example, realm = "atmos" for many variables or realm = "landice land" for a few variables including surface snow amount where land.
4. an attribute assigned a value or multiple values found in a CV (with no restriction on what could be included), unchecked
For example, source_type = "AGCM CHEM AER".

The values defined in the [CMIP6_CVs/CMIP6_required_global_attributes.json](https://github.com/WCRP-CMIP/CMIP6_CVs/blob/master/CMIP6_required_global_attributes.json) file can be broken down into the 4 subgroups above:

1 single value, no registry/CVs, unchecked
```python
        "Conventions",
        "creation_date",
        "forcing_index",
        "further_info_url",  # Recommended, provided by ES-Docs
        "initialization_index",
        "license",  # Recommended
        "physics_index",
        "product",
        "realization_index",
        "tracking_id",  # CMOR written
        "variant_label"
```
2 single value, registry/CVs, checked
```python
        "activity_id",  #CVs
        "data_specs_version",  # table header
        "experiment",  #CVs
        "experiment_id",  #CVs
        "frequency",  #CVs
        "grid",  #CVs
        "grid_label",  #CVs
        "institution",  #CVs
        "institution_id",  #CVs
        "mip_era",  #CVs
        "nominal_resolution",  #CVs
        "source",  #CVs
        "source_id",  #CVs
        "sub_experiment",  #CVs
        "sub_experiment_id",  #CVs
        "table_id",  # table header
        "variable_id",  # table variable
```
3 multiple values, registry/CVs, checked
```python
        "realm",  #CVs
```
3 multiple values, registry/CVs, unchecked
```python
        "source_type",
```

To implement the above, we would need to augment the structure of the `CMIP6_required_global_attributes.json` to provide this information programmatically

@taylor13 @matthew-mizielinski does the above seem correct to you?

**Comment by durack1** [comment]
And @mauzey1 to make sense of the above, the group 2 entries are defined to a single (registered) entry, to ensure that for example in the case that a query is raised against a model that is registered against multiple institution_id entries, it is obviously where queries should be directed (to the single institution_id listed in the file, and noted in the ESGF publication database)

**Comment by matthew-mizielinski** [comment]
Apologies for missing this.

> @taylor13 @matthew-mizielinski does the above seem correct to you?

I think `grid` is a free-form text field, we use it to add a string describing the grid in use, I've put a selection of the ones we have used in CMIP6 from UKESM1-0-LL and HadGEM3-GC31-LL (similar for HadGEM3-GC31-MM, but with more points)

```
grid = Native N96 grid (U points); 192 x 144 longitude/latitude
grid = Native N96 grid (UV points); 192 x 145 longitude/latitude
grid = Native N96 grid (UV points); Zonal mean, 145 latitude
grid = Native N96 grid (V points); 192 x 145 longitude/latitude
grid = Native N96 grid; 192 x 144 longitude/latitude
grid = Native N96 grid; Global mean
grid = Native N96 grid; Zonal mean, 144 latitude
grid = Native eORCA1 tripolar primarily 1 deg with meridional refinement down to 1/3 degree in the tropics; 360 x 330 longitude/latitude
grid = Native eORCA1 tripolar primarily 1 deg with meridional refinement down to 1/3 degree in the tropics; Global mean
grid = Native eORCA1 tripolar primarily 1 deg with meridional refinement down to 1/3 degree in the tropics; Quasi-zonal mean, 330 latitude
```

**Comment by durack1** [comment]
@matthew-mizielinski thanks, I have corrected the above https://github.com/PCMDI/cmor/issues/628#issuecomment-1085241113. You're right, `grid` is a freeform field, `grid_label` is the CV-controlled attribute

**Comment by mauzey1** [comment]
@durack1 So we want to add information to the CV file that says which attribute can have multiple values listed?  This will require changes to CMOR to read but the CV file should still be usable by older CMOR versions, which should ignore this new data.

`realm` and `source_type`aren't listed in the registered entries but `institution_id` and `region` are.  If we encountered attributes in registered entries with bracketed list that contain multiple entries such as [this](https://github.com/PCMDI/cmip6-cmor-tables/blob/196f265c1c2f92db15da9d6f5d21e15010b281b0/Tables/CMIP6_CV.json#L510-L513), should we report an error if this attribute is missing from the dataset rather than trying to fill it with the value list?  Maybe display an error message similar to the one below.

```
!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: The registered CV attribute "institution_id" has multiple values defined in CMIP6_CV.json.
! Please use one of the following values for this attribute.
!            * UCI
!            * NCAR
! 
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```

If there is only one value listed in the bracketed list, then maybe we could use that one value for the dataset just like the other single-value attributes.

**Comment by durack1** [comment]
@mauzey1 exactly (as you describe https://github.com/PCMDI/cmor/issues/628#issuecomment-1092279751), what would be the best way to do this, add an entry to the [CMIP6_CV.json](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_CV.json)?

**Comment by mauzey1** [comment]
@durack1 Maybe have a section next to `required_global_attributes` called `multiple_value_attributes` listing the attributes that can have a list of values.

As for the values that aren't listed as having multiple values but have multiple values in the registered content, should I create the kind of error that I described in https://github.com/PCMDI/cmor/issues/628#issuecomment-1092279751?  Otherwise, CMOR will just use the missing attribute error that it currently uses.

**Comment by mauzey1** [comment]
@durack1 

I have made changes similar to what I have described in https://github.com/PCMDI/cmor/issues/628#issuecomment-1092279751 where single-value bracketed values will be treated as single values for attributes, but multiple values will raise an error asking the user to select a value.  It will only use these values if their attribute is listed in `required_global_attributes`.

Here's an example of an error message.
```
C Traceback:
! In function: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: The registered CV attribute "source_variables" has multiple values 
! defined in "./input4MIPs-cmor-tables/Tables/input4MIPs_CV.json"
! Please select one from the entry source_id.PCMDI-AMIP-1-2-0.source_variables.
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```

`source_variables` is not in `required_global_attributes`, I'm just using it as an example

I will push these changes into the testing branch.

**Comment by durack1** [comment]
@mauzey1 great, thanks for pushing on this!

**Comment by durack1** [comment]
@mauzey1 apologies for the delay in getting back to this. Now we also have #656 created, is it just this and the license template to be finalized before we can get 3.7.0 released and tagged?

I have to generate an updated PCMDI-AMIP dataset update soon, so would ideally like to do this with CMOR 3.7.0 if that was possible in the coming week(s)?

@matthew-mizielinski @taylor13 ping

**Comment by durack1** [comment]
@mauzey1 just circling around on this - the new build `cmor 3.6.1.2022.08.02.00.28.g4871d9c  py310h4b0e41b_0          PCMDI/label/nightly/linux-64` is looking great, but I have one last tweak that I am trying to ascertain.

When I declare a new license string, I am getting hit with the error:
```python
C Traceback:
In function: _CV_checkGblAttributes
! called from: cmor_setGblAttr
! called from: cmor_write
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Warning: The registered CV attribute "license" as defined as "AMIP boundary condition data produced by PCMDI is licensed under a Creative Commons Attribution 4.0 International License (CC BY 4.0; https://creativecommons.org/licenses/by/4.0). Consult https://pcmdi.llnl.gov/CMIP6/TermsOfUse for terms of use governing input4MIPs output, including citation requirements and proper acknowledgment. Further information about this data, including some limitations, can be found via the further_info_url (recorded as a global attribute in this file). The data producers and data providers make no warranty, either express or implied, including, but not limited to, warranties of merchantability and fitness for a particular purpose. All liabilities arising from the supply of the information (including any liability arising in negligence) are excluded to the fullest extent permitted by law" will be replaced with 
! "AMIP boundary condition data produced by PCMDI is licensed under a Creative Commons Attribution-[NonCommerci
!
!!!!!!!!!!!!!!!!!!!!!!!!!
```

The **Creative Commons Attribution-[NonCommerci..** is not in any of my input files (that I can find), is this a hard-coded error?

**Comment by mauzey1** [comment]
@durack1 Which CV file are you using?  What does your user input file look like?

**Comment by durack1** [comment]
@mauzey1 this turned out to be an issue with one of two user_input.json files, in which I got a little confused which one was being used - user error, sorry!

Good news, I have validated that the slimmed down `user_input.json` files found at [`amipbcs` user_input files](https://github.com/PCMDI/amipbcs/pull/29/files) write the output as expected with the nightly/test build - this outstanding issue is closed - on with the 3.7.0 release!~!
